colocate string and tuple data

toshok opened the pull request:
There's a lot of commits here, but most of them are squashable.  The crux of the change is both BoxedTuple and BoxedString have specialized `operator new`s that calculate the size of the Boxed structure as well as any additional data (Box*'s for the BoxedTuple, chars for BoxedString).

This should reduce the number of calls to our allocator by at least a half for these types.  In order to keep more of the BoxedString code the same, I added `pyston::String` which can mostly be described as "an llvm::StringRef that exposes a std::string api".  One difference is that `pyston::String`'s data can be owned (malloced/freed by pyston::String), which happens using `operator+`, etc.  It's normal mode is unowned (its use in BoxedString).

Also some effort goes to calculating the number of elements (tuple elements, characters) in the result before doing the allocation for the box, so we don't have to realloc.  This means we need non-ctor allocation paths (we already had them for BoxedString, in the boxString* family).  For BoxedTuple, I added static BoxedTuple::Create() overloads.

performance numbers initially were pretty crazy (-10% for interp2), but changes on master have brought them down to a more modest level (I haven't investigated to see which change(s) on master caused this.)  Current comparison shows:

```
              pyston (calibration)                      :    1.0s baseline: 1.0 (+0.0%)
              pyston interp2.py                         :    3.8s baseline: 3.9 (-2.5%)
              pyston raytrace.py                        :    6.4s baseline: 6.3 (+1.0%)
              pyston nbody.py                           :    8.7s baseline: 8.5 (+2.7%)
              pyston fannkuch.py                        :    6.6s baseline: 6.6 (+0.4%)
              pyston chaos.py                           :   20.1s baseline: 19.9 (+1.4%)
              pyston spectral_norm.py                   :   21.1s baseline: 22.5 (-6.2%)
              pyston fasta.py                           :    4.9s baseline: 5.3 (-6.9%)
              pyston pidigits.py                        :    5.1s baseline: 5.0 (+1.6%)
              pyston richards.py                        :    1.9s baseline: 1.8 (+1.0%)
              pyston deltablue.py                       :    2.0s baseline: 2.0 (-0.2%)
              pyston (geomean-0b9f)                     :    6.0s baseline: 6.0 (-0.8%)
```

kmod commented on src/capi/object.cpp:
```diff
@@ -146,7 +146,7 @@ extern "C" PyObject* PyObject_SelfIter(PyObject* obj) noexcept {
 extern "C" int PyObject_GenericSetAttr(PyObject* obj, PyObject* name, PyObject* value) noexcept {
     try {
         if (value == NULL)
-            delattrGeneric(obj, static_cast<BoxedString*>(name)->s, NULL);
```
It'd be great to have functions start taking llvm::StringRef or Pyston::String rather than allocating a fresh std::string (doesn't have to be in this PR).

toshok replied:
agreed.  we could also support llvm::Twine's, though those aren't as handy if callers aren't calling with concatenated strings.

kmod commented on src/runtime/types.h:
```diff
@@ -435,14 +455,63 @@ class BoxedList : public Box {
+        return rtn;
+    }
+
+    static BoxedTuple* Create(int64_t size) { return new (size) BoxedTuple(size); }
```
We've usually been using camelCase for function names (ie this would be `create`).

kmod commented on src/core/string.h:
```diff
@@ -0,0 +1,231 @@
+
+class String {
+public:
+    explicit String(const std::string& s, char* _ptr = NULL) : ptr(_ptr), own(_ptr == NULL) {
```
Can we get a comment about the ownership strategy?  I was somewhat confused by the way this works, since I was assuming it would work similarly to llvm::StringRef in the non-owning case (ie would just reference the original data rather than copy it).

toshok replied:
I found and fixed the cases where we were creating temporaries (operator+, operator<<), and switched those instances to using llvm::Twines and llvm::raw_string_ostream.  The latter lets us use llvm::StringRef directly, and the former delays the concatenation until the place where the value is used (in the new `boxStringTwine` method).  Now there are no cases where pyston::String owns the buffer, so it shrinks by a bool, the ctors get simpler, and there's no dtor.

toshok commented:
this will definitely need squashing, but I think everything's been addressed/rebased off master.

toshok commented:
hrm, actually after the last rebase the django migrate test is crashing... ugh.

toshok commented:
actually it appears to be the switch to -x by default that's crashing the django migrate test.

undingen commented:
I could reproduce the libpypa bug with django on current head, I will look into it.

kmod commented on src/runtime/util.cpp:
```diff
@@ -39,32 +39,32 @@ void parseSlice(BoxedSlice* slice, int size, i64* out_start, i64* out_stop, i64*
     int64_t istep = 1;
 
-    if (isSubclass(step->cls, int_cls)) {
+    if (step->cls != none_cls && (step->cls == int_cls || isSubclass(step->cls, int_cls))) {
```
?

toshok replied:
ah I'd added that just to reduce calls to isSubclass (which were rather hot in `make perf` output while I was working on this.)  given @undingen's changes this should be unnecessary.

kmod commented on test/tests/weakref1.py:
```diff
@@ -1,3 +1,4 @@
+# skip-if: '-n' not in EXTRA_JIT_ARGS
```
Are these tests working now with the changes I made recently?

toshok replied:
I thought I'd removed those.  they're working.

kmod commented on src/core/string.h:
```diff
@@ -0,0 +1,169 @@
+
+class String {
+public:
+    explicit String(const std::string& s, char* _ptr) : ptr(_ptr) {
```
I still think it's a weird mix of concerns that the storage pointer must be provided by the caller, but the constructor takes care of the copying.  It kind of feels like we just took some of the BoxedString functionality and put it in another class... I guess I'm glad we tried to make String reusable, but I just have a hard time seeing where this would be used other than BoxedString.

What if we have the BoxedString constructor do this copying, and give it a llvm_str function that returns a llvm::StringRef?

toshok replied:
I can definitely move the copying from pyston::String to BoxedString.  I'll look at how large the changes are to removing pyston::String and replacing it with the llvm alternatives.

kmod commented on src/runtime/str.cpp:
```diff
@@ -46,16 +46,33 @@ extern "C" PyObject* string_splitlines(PyStringObject* self, PyObject* args) noe
+    } else {
+        memset(storage(), 0, n + 1);
+    }
+    gc::registerGCManagedBytes(cls->tp_basicsize + this->s.size() + 1);
```
I don't think we have to call this any more; it was there since the gc wouldn't see the allocation that the s constructor (as a std::string) would do.  I think this should get seen now by the allocation for the object in operator new?

toshok replied:
yup, the various arena `alloc` functions already register.  so with these lines we double count string memory.

kmod commented on src/runtime/import.cpp:
```diff
@@ -441,9 +441,8 @@ static bool loadNext(Box* mod, Box* altmod, std::string& name, std::string& buf,
 
 static void ensureFromlist(Box* module, Box* fromlist, std::string& buf, bool recursive);
-Box* importModuleLevel(std::string* name, Box* globals, Box* from_imports, int level) {
+Box* importModuleLevel(std::string& name, Box* globals, Box* from_imports, int level) {
```
should probably be a const reference?

kmod commented on src/runtime/import.cpp:
```diff
@@ -453,7 +452,7 @@ Box* importModuleLevel(std::string* name, Box* globals, Box* from_imports, int l
         return NULL;
 
-    std::string _name = *name;
+    std::string _name = name;
```
heh let's kill this, not sure why we have it.

toshok replied:
this one is necessary (we need the copy) because loadNext actually modifies _name.

kmod commented on src/runtime/types.h:
```diff
@@ -371,14 +375,31 @@ class BoxedBool : public BoxedInt {
-    std::string s;
+    llvm::StringRef s;
+
+    char* storage() { return (char*)this + cls->tp_basicsize; }
```
Hmm so some places call storage(), and some look at `->s.data()`.  I guess if we're going to store the pointer we should prefer to use it?  Maybe we could do something like have a private function that calculates the storage offset, and a public one that just returns s.data?

kmod commented on src/runtime/types.h:
```diff
@@ -371,14 +375,31 @@ class BoxedBool : public BoxedInt {
+    }
 
+    void* operator new(size_t size, BoxedClass* cls, size_t ssize) __attribute__((visibility("default"))) {
+        // XXX this is likely broken...
```
(why?)

toshok replied:
ah, because I hadn't verified that `tp_itemsize == sizeof(char)` in cpython.  just did that, and removing comment.

kmod commented on src/runtime/util.cpp:
```diff
@@ -46,25 +46,25 @@ void parseSlice(BoxedSlice* slice, int size, i64* out_start, i64* out_stop, i64*
-        if (istart < 0)
-            istart = size + istart;
-    } else {
+    if (start->cls == none_cls || (start->cls != int_cls && !isSubclass(start->cls, int_cls))) {
```
Do we need these changes?  it looks like it still has the same behavior but I hate touching the slice parsing code...

toshok replied:
I'd like to spend some time with them in the future (once verifying that they're still needed after the `isSubclass` work), but definitely not something needed for this change.

kmod commented:
lgtm!  lots of opportunities to reduce string copies after this :)

could you squash before pushing?  also if it's not too hard, implementing _PyTuple_Resize would be great as well (I needed an implementation for my branch but it will break with this change).

toshok commented:
I've squashed everything except the final `->storage()` change.  will squash again before merging.

we can't directly use `->s.data()` everywhere (at least without `const_cast`ing it), so I've made the following changes:

1. `::storage()` is private, and only used in ctors when initializing the `llvm::StringRef`
2. added `::data()` which returns `s.data()`, but casted to `char*`
3. added `::size()` for consistency with `::data()` - just returns `s.size()`
